tools/buildgen/extract_metadata_from_bazel_xml.py

```diff
@@ -245,98 +242,171 @@ def _external_dep_name_from_bazel_dependency(bazel_dep):
         return None
 
 
-def _expand_intermediate_deps(target_dict, public_dep_names, bazel_rules):
-    # Some of the libraries defined by bazel won't be exposed in build.yaml
-    # We call these "intermediate" dependencies. This method expands
-    # the intermediate deps for given target (populates library's
-    # headers, sources and dicts as if the intermediate dependency never existed)
-
-    # use this dictionary to translate from bazel labels to dep names
-    bazel_label_to_dep_name = {}
-    for dep_name in public_dep_names:
-        bazel_label_to_dep_name[_get_bazel_label(dep_name)] = dep_name
-
-    target_name = target_dict['name']
-    bazel_deps = target_dict['_DEPS_BAZEL']
-
-    # initial values
-    public_headers = set(target_dict['_PUBLIC_HEADERS_BAZEL'])
-    headers = set(target_dict['_HEADERS_BAZEL'])
-    src = set(target_dict['_SRC_BAZEL'])
-    deps = set()
+def _deduplicate_append(target: List[str], pending: List[str]) -> None:
+    """Append the list without duplicated items in place.
+    
+    Expected to be used to create a order-stable dependency list.
+    """
+    seen = set(target)
+    for dep in pending:
+        if dep not in seen:
+            target.append(dep)
+
+
+def _compute_transitive_metadata(
+        rule_name: str, bazel_rules: Any,
+        bazel_label_to_dep_name: Dict[str, str]) -> None:
+    """Computes the final build metadata for Bazel target with rule_name.
+
+    The dependencies that will appear on the deps list are:
+
+    * Public build targets including binaries and tests;
+    * External targets, like absl, re2.
+
+    All other intermediate dependencies will be merged, which means their
+    source file, headers, etc. will be collected into one build target. This
+    step of processing will greatly reduce the complexity of the generated
+    CMake build files.
+
+    The final build metadata are:
+    * _TRANSITIVE_DEPS: all the transitive dependencies including intermediate
+                        targets;
+    * _COLLAPSED_DEPS:  dependencies that fits our requirement above, and it
```

Comment:
are you 100% sure that this is correct?  Imagine a hypothetical situation.
1. grpc++  depends on grpc 
2. both grpc++ and grpc depend on gpr (and they actually both invoke a functions from libgpr)

I think when invoking a final linker command, it should be:
`link --output grpc libgpr.a`
`link --output grpc++ libgrpc.a libgpr.a`
(note that grpc++ gets linked against both libgrpc.a and libgpr.a because it actually does invoke functions from both libraries).

Can you actually break a build by only invoking `link --output grpc++ libgrpc.a` (That is what our Makefile is going to do if you only include `grpc` in `deps` in build.yaml)? If so, than the approach you're proposing is actually not correct, it just happens to work for our build (basically by coincidence) - that also means that it could break unexpectedly and we really don't want sudden breakages like this.

Reply:
I did some experiment around the `make` artifacts. Since `Makefile` are now Core-only, I just tested if Core works. Here is a simple code that prints channelz stats, which underlying invokes methods from `gpr`/`abseil`. If the linking has failed or has missing component, it should fail:

```c
#include <stdio.h>
#include <grpc/grpc.h>
#include <grpc/grpc.h>
int main(int argc, char** argv) {
  grpc_init();
  printf("%s\n", grpc_channelz_get_top_channels(0));
  grpc_shutdown();
  return 0;
}
```
```
$ gcc simple.c -lgrpc && ./a.out
{"end":true}
```

The output seems fine, and the binary isn't crashing. TBH, I think I'm weak on how `ld` functions, but if you have a test that can proof this build is broken, I can try it out. So far, I haven't seen evidence that that this new linking method won't work.

I guess your worry might came true if we strip symbols aggressively for each of the `.a`. CC @gnossen can you help us?

Reply:
I'll look into this and if I find a problem, I'll try to provide and example.